cli/azd/pkg/auth/guard.go

```diff
@@ -0,0 +1,19 @@
+package auth
+
+import "context"
+
+type EnsureLoginGuard struct{}
+
+func NewEnsureLoginGuard(manager *Manager, ctx context.Context) (EnsureLoginGuard, error) {
+	cred, err := manager.CredentialForCurrentUser(ctx, nil)
+	if err != nil {
+		return EnsureLoginGuard{}, err
+	}
+
+	_, err = EnsureLoggedInCredential(ctx, cred)
+	if err != nil {
+		return EnsureLoginGuard{}, err
+	}
+
+	return EnsureLoginGuard{}, nil
```

Comment:
why do we want to return empty struct?  

Isn't this actually a `EnsureCredentialLoggedIn` which should return the credential or a true/false?

Reply:
The constructor returns `T` if success, `err` if failure.

Maybe we need to rename this to `LoggedinGuard` instead of `EnsureLoginGuard`

Reply:
What I mean is, we are adding `T` as a dependency for actions like `Init`. But the action does not care about `T` beyond the fact that it can be created...

Then, creating the `T` depends on creating `manager *Manager`, because `T` will get the `crededential` from the manager and check that it is logged in.

So, what an action like `init` really needs is to know that there is a logged in credential. And we are using the logic to create a `T` dependency as the way to know this.  Which I don't think is convenient.

If what the action (like init), really needs is a `logged in credential`, then make the `T` to be either the credential, or an struct that contains the `logged in credential`.  Otherwise, it's odd to see a dependency to be a dependency but to actually be ignored, right... (why is it a dependency then...)

Reply:
It's used for creating a side-effect as a `Guard`. I don't know what we want to pass back to `init`... It doesn't actually need the `credential`. It simply needs the `Guard` to fail-fast. Removing the code would have the command still working. It is simply added to create early validation.

Reply:
This feels like a good implementation of this pattern to me.  It feels very idiomatic to use the discard as we do here https://github.com/Azure/azure-dev/pull/1570/files#diff-cd0cee372992ad11f4b9477111cf0e3025989cb13454ec6a2d3a140f0b8c81bbR112 to express that we only care about the side-effects of the component being injected, similar to when we import packages to just their `init()` methods.  I think `LoggedInGuard` is a good name.  I like the `Guard` suffix to express this pattern.

Reply:
Can we use a more general pattern here?  like a `factory` or `builder` to produce the `initAction` ?
Then, any number of `guards` or pre-conditions are implemented within the `object producer` and not as an input for the actual `object`.

Reply:
I'm going to land this for now so that we correctly fix `login` errors for `azd up` for the next release.

I'd be happy to talk to you @vhvb1989 to figure out the design here. I suspect it comes down to minute differences that won't actually impact us. And if we need to make changes here, it'd be a few lines of code that we could easily adjust to.

> Then, any number of guards or pre-conditions are implemented within the object producer and not as an input for the actual object.

This is actually the current behavior. You can examine this by noticing that you are able to construct an `action` with the default `LoggedInGuard` struct and it works correctly. The `producer` is the one that is performing the pre-validation.

Abstractly, adding the guard to the constructor is the same as defining a dependency. However, the dependency is only used for side-effects and not actually to be used as a component. This is no different than how constructing a normal struct would work -- not all options are stored as instance variables in the constructed struct.